### PR TITLE
feat(treesitter): allow injections to be configured through directives

### PR DESCRIPTION
**Pull request by steelsojka**
This features makes the injection query api more generalized through directive configuration. Directives can now be used to set injection properties such as node ranges, language, and whether ranges should be combined. Here are some examples:

Basic example using `set` directive.
```scheme
((comment) (#set! "language" "jsdoc"))
```

Complex example using a custom directive that resolves the language based on a buffer variable set by the user:

```lua
vim.treesitter.add_directive("css-preprocessor!", function(match, _, bufnr, pred, metadata)
  local success, lang = pcall(vim.api.nvim_buf_get_var, bufnr, "style_lang")

  metadata.language = success and lang or "css"
end)
```

```scheme
(style_element
  (raw_text) (#css-preprocessor!))
```

The above example could be used for changing which type of css preprocessor is used to parse style tags based on what the user has configured (SASS, less, stylus, etc...).

This api will allow complex injections through directives instead of relying on core changes in neovim. For example "fenced" language injection can be implemented in nvim-treesitter without requiring any changes to neovim.

### CODE REVIEW

**Comment by theHamsta** [comment]
Could you have a look onto https://github.com/nvim-treesitter/nvim-treesitter/pull/995 I guess you and @tjdevries are tackling similar problems

**Comment by steelsojka** [comment]
@theHamsta They are both similar in how it gives some control to the user, but I don't think there is any overlap. The key thing with this PR is allowing advanced processing for injection ranges and language resolution. For example, this is a precursor for work in injecting languages into a block of text instead of just a single nodes range.

```
/**
  * Some jsdoc description
  * 
  * @example
  * const someVar = "blorg";
  */
```

We could highlight that js code as actually javascript instead of having it handled by the jsdoc parser. This PR allows us to do that without having to do any changes in neovim.

File: runtime/lua/vim/treesitter/languagetree.lua
Comment (elianiva) [inline runtime/lua/vim/treesitter/languagetree.lua:305]: just a nitpick

```suggestion
      -- This allows more advanced processing for determining ranges and language resolution.
```

**Review by vigoux — APPROVED** [review]

**Comment by steelsojka** [comment]
@vigoux Is this good to be merged?

File: test/functional/treesitter/parser_spec.lua
Comment (tjdevries) [inline test/functional/treesitter/parser_spec.lua:576]: Is this no longer possible to do, is that why the test is removed?

File: test/functional/treesitter/parser_spec.lua
Comment (tjdevries) [inline test/functional/treesitter/parser_spec.lua:512]: It'd be good to add some more information / documentation about this I think.

Comment (steelsojka) [reply test/functional/treesitter/parser_spec.lua:512]: I can add some info to the vim docs

Comment (steelsojka) [reply test/functional/treesitter/parser_spec.lua:512]: Done!

File: test/functional/treesitter/parser_spec.lua
Comment (steelsojka) [reply test/functional/treesitter/parser_spec.lua:576]: The old implementation had the concept of storing data pure node in the metadata strictly for the way offset was storing data. This was removed since we only need the metadata for the match instead of on a capture basis. So this is no longer used and was removed.

**Review by vigoux — APPROVED** [review]

File: runtime/lua/vim/treesitter/languagetree.lua
Comment (DerekStride) [inline runtime/lua/vim/treesitter/languagetree.lua:320]: I think this is great work allowing the directive's to have control over the language and content of injected nodes. I suggested a similar idea here https://github.com/neovim/neovim/pull/14152 but without the ability to modify the content.

Merging this would enable a functioning `downcase!` / `lowercase!` directive in my PR as an example.

**Review by DerekStride — APPROVED** [review]

**Comment by steelsojka** [comment]
@vigoux @tjdevries Is this good to merge? 😉

**Comment by steelsojka** [comment]
@vigoux @tjdevries Any way we can get this merged in? This will add a lot of flexibility for the language injection queries.

**Comment by theHamsta** [comment]
It can still be extended later to folds/indents for a unified solution

**Comment by steelsojka** [comment]
@theHamsta I agree, to an extent... I don't think there is much overlap between the implementations. These properties only make sense in the context of injections (maybe except for the ranges).

**Comment by tjdevries** [comment]
While we're making some changes here, what are you thoughts on changing the name from just being `queries` as the key to `injections` or `injeciton_queries` or ` queries = { injections = ... } }` ?

Just wondering.

I think I'll merge if no one has objections since we got approvals earlier after we chat about this question.

**Comment by steelsojka** [comment]
@tjdevries  I can make that change. I think injection_queries would work pretty good.

File: runtime/doc/treesitter.txt
Comment (tjdevries) [inline runtime/doc/treesitter.txt:429]: this should be changed to `injection_queries` then?

File: runtime/lua/vim/treesitter/languagetree.lua
Comment (tjdevries) [inline runtime/lua/vim/treesitter/languagetree.lua:32]: only weird thing now is that we have `injection_queries` (as the key) and `injections` (as the file) which I'm not sure if that will be confusing.

Do you have any thoughts about that?

File: runtime/doc/treesitter.txt
Comment (steelsojka) [reply runtime/doc/treesitter.txt:429]: yep. Missed that one.

File: runtime/lua/vim/treesitter/languagetree.lua
Comment (steelsojka) [reply runtime/lua/vim/treesitter/languagetree.lua]: changed!

File: runtime/lua/vim/treesitter/languagetree.lua
Comment (tjdevries) [inline runtime/lua/vim/treesitter/languagetree.lua:22]: What do you think about doing someting like `assert(not opts.queries, "'queries' is no longer supported. Use 'injections' now")` or something like that?

**Comment by tjdevries** [comment]
And yes, I like just having `injections` btw, that makes it a lot simpler to reason about.

File: runtime/lua/vim/treesitter/languagetree.lua
Comment (steelsojka) [reply runtime/lua/vim/treesitter/languagetree.lua:22]: I'm good with that

File: runtime/lua/vim/treesitter/languagetree.lua
Comment (vigoux) [inline runtime/lua/vim/treesitter/languagetree.lua:22]: Maybe using an assert here is a bit to harsh ?
Just saying that to the user, and falling back would be better IMO.

File: runtime/lua/vim/treesitter/languagetree.lua
Comment (vigoux) [inline runtime/lua/vim/treesitter/languagetree.lua:326]: If `metadate.combined` is `nil` or `false`, this'll be the same when using it in `if`s so no need to check here.

File: runtime/lua/vim/treesitter/languagetree.lua
Comment (steelsojka) [reply runtime/lua/vim/treesitter/languagetree.lua:22]: So `print`?

Comment (steelsojka) [reply runtime/lua/vim/treesitter/languagetree.lua:22]: done

**Review by vigoux — APPROVED** [review]

**Comment by steelsojka** [comment]
@vigoux I'm thinking of expanded this to the other query types as well. So you could resolve the capture of highlights with a predicate `(#set! "capture" "number")` or some more complex logic. It's trivial to add it to highlights... do you want me to add it to this PR or another one after this one?

**Comment by DerekStride** [comment]
> do you want me to add it to this PR or another one after this one?

FWIW I think a separate PR makes the most sense

**Comment by theHamsta** [comment]
@steelsojka I think that such preprocessor for highlighting make much more sense than `is-local? @parameter "parameter"` because the Lua function could just determine what it is instead of going throug `is-foo?` `is-baz?`

**Comment by steelsojka** [comment]
@vigoux I'll do another PR for any highlight related stuff. This one should still be good to go!

**Pull request merged**
